File: src/utils/logger.rb
require 'time'

module PTD
  module Utils
    class Logger
      LEVELS = {
        debug: 0,
        info: 1,
        warn: 2,
        error: 3,
        fatal: 4
      }.freeze

      COLORS = {
        debug: "\e[36m",  # Cyan
        info: "\e[32m",   # Green
        warn: "\e[33m",   # Yellow
        error: "\e[31m",  # Red
        fatal: "\e[35m",  # Magenta
        reset: "\e[0m"
      }.freeze

      attr_accessor :level, :use_colors

      def initialize(level: :info, verbose: false, use_colors: true, output: $stdout)
        @level = verbose ? :debug : level
        @use_colors = use_colors && output.tty?
        @output = output
        @mutex = Mutex.new
      end

      def debug(message, metadata = {})
        log(:debug, message, metadata)
      end

      def info(message, metadata = {})
        log(:info, message, metadata)
      end

      def warn(message, metadata = {})
        log(:warn, message, metadata)
      end

      def error(message, metadata = {})
        log(:error, message, metadata)
      end

      def fatal(message, metadata = {})
        log(:fatal, message, metadata)
      end

      def with_timing(message)
        start_time = Time.now
        info("Starting: #{message}")

        result = yield

        elapsed = Time.now - start_time
        info("Completed: #{message} (#{format_duration(elapsed)})")

        result
      rescue StandardError => e
        elapsed = Time.now - start_time
        error("Failed: #{message} (#{format_duration(elapsed)}) - #{e.message}")
        raise
      end

      def progress(current, total, message = 'Progress')
        percentage = (current.to_f / total * 100).round(1)
        bar_length = 30
        filled = (bar_length * (current.to_f / total)).round

        bar = ('█' * filled) + ('░' * (bar_length - filled))

        @mutex.synchronize do
          @output.print "\r#{message}: [#{bar}] #{percentage}% (#{current}/#{total})"
          @output.print "\n" if current >= total
        end
      end

      private

      def log(severity, message, metadata = {})
        return if LEVELS[severity] < LEVELS[@level]

        timestamp = Time.now.iso8601
        formatted_message = format_message(severity, timestamp, message, metadata)

        @mutex.synchronize do
          @output.puts formatted_message
        end
      end

      def format_message(severity, timestamp, message, metadata)
        severity_str = severity.to_s.upcase.ljust(5)

        if @use_colors
          severity_color = COLORS[severity]
          reset_color = COLORS[:reset]
          formatted = "#{severity_color}[#{timestamp}] #{severity_str}#{reset_color} | #{message}"
        else
          formatted = "[#{timestamp}] #{severity_str} | #{message}"
        end

        formatted += " | #{format_metadata(metadata)}" unless metadata.empty?

        formatted
      end

      def format_metadata(metadata)
        metadata.map { |k, v| "#{k}=#{v.inspect}" }.join(' ')
      end

      def format_duration(seconds)
        if seconds < 1
          "#{(seconds * 1000).round(2)}ms"
        elsif seconds < 60
          "#{seconds.round(2)}s"
        else
          minutes = (seconds / 60).floor
          secs = (seconds % 60).round
          "#{minutes}m #{secs}s"
        end
      end
    end

    class FileLogger < Logger
      def initialize(filename, **options)
        file = File.open(filename, 'a')
        file.sync = true
        super(**options.merge(output: file, use_colors: false))

        at_exit { file.close unless file.closed? }
      end
    end

    class MultiLogger
      def initialize(*loggers)
        @loggers = loggers
      end

      %i[debug info warn error fatal].each do |level|
        define_method(level) do |message, metadata = {}|
          @loggers.each { |logger| logger.send(level, message, metadata) }
        end
      end

      def with_timing(message, &block)
        @loggers.first.with_timing(message, &block)
      end

      def progress(current, total, message = 'Progress')
        @loggers.each { |logger| logger.progress(current, total, message) }
      end
    end
  end
end

File: src/utils/logger.rs
use chrono::Local;
use colored::*;
use indicatif::{ProgressBar, ProgressStyle};
use std::io::Write;
use std::sync::Mutex;
use std::time::{Duration, Instant};

#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord)]
#[allow(dead_code)]
pub enum LogLevel {
    Debug = 0,
    Info = 1,
    Warn = 2,
    Error = 3,
    Fatal = 4,
}

pub struct Logger {
    level: LogLevel,
    use_colors: bool,
    output: Mutex<Box<dyn Write + Send>>,
}

#[allow(dead_code)]
impl Logger {
    pub fn new(level: LogLevel) -> Self {
        Self {
            level,
            use_colors: atty::is(atty::Stream::Stdout),
            output: Mutex::new(Box::new(std::io::stdout())),
        }
    }

    pub fn new_with_options(level: LogLevel, use_colors: bool) -> Self {
        Self {
            level,
            use_colors,
            output: Mutex::new(Box::new(std::io::stdout())),
        }
    }

    pub fn debug(&self, message: &str) {
        self.log(LogLevel::Debug, message);
    }

    pub fn info(&self, message: &str) {
        self.log(LogLevel::Info, message);
    }

    pub fn warn(&self, message: &str) {
        self.log(LogLevel::Warn, message);
    }

    pub fn error(&self, message: &str) {
        self.log(LogLevel::Error, message);
    }

    pub fn fatal(&self, message: &str) {
        self.log(LogLevel::Fatal, message);
    }

    fn log(&self, severity: LogLevel, message: &str) {
        if severity < self.level {
            return;
        }

        let timestamp = Local::now().format("%Y-%m-%dT%H:%M:%S%.3f");
        let severity_str = format!("{:?}", severity).to_uppercase();

        let formatted = if self.use_colors {
            let colored_severity = match severity {
                LogLevel::Debug => severity_str.cyan(),
                LogLevel::Info => severity_str.green(),
                LogLevel::Warn => severity_str.yellow(),
                LogLevel::Error => severity_str.red(),
                LogLevel::Fatal => severity_str.magenta(),
            };
            format!("[{}] {} | {}", timestamp, colored_severity, message)
        } else {
            format!("[{}] {:5} | {}", timestamp, severity_str, message)
        };

        let mut output = self.output.lock().unwrap();
        writeln!(output, "{}", formatted).unwrap();
    }

    pub fn with_timing<F, R>(&self, message: &str, f: F) -> R
    where
        F: FnOnce() -> R,
    {
        let start = Instant::now();
        self.info(&format!("Starting: {}", message));

        let result = f();

        let elapsed = start.elapsed();
        self.info(&format!(
            "Completed: {} ({})",
            message,
            format_duration(elapsed)
        ));

        result
    }

    pub fn progress(&self, current: usize, total: usize, message: &str) {
        let _percentage = (current as f64 / total as f64 * 100.0) as u64;
        let pb = ProgressBar::new(total as u64);

        pb.set_style(
            ProgressStyle::default_bar()
                .template(&format!(
                    "{}: [{{bar:30}}] {{percent}}% ({{pos}}/{{len}})",
                    message
                ))
                .unwrap()
                .progress_chars("█▉▊▋▌▍▎▏  "),
        );

        pb.set_position(current as u64);

        if current >= total {
            pb.finish_and_clear();
            println!();
        }
    }
}

impl Default for Logger {
    fn default() -> Self {
        Self::new(LogLevel::Info)
    }
}

#[allow(dead_code)]
fn format_duration(d: Duration) -> String {
    if d.as_secs() > 60 {
        let minutes = d.as_secs() / 60;
        let seconds = d.as_secs() % 60;
        format!("{}m {}s", minutes, seconds)
    } else if d.as_secs() > 0 {
        format!("{:.2}s", d.as_secs_f64())
    } else {
        format!("{:.2}ms", d.as_millis() as f64)
    }
}

#[allow(dead_code)]
pub struct FileLogger {
    logger: Logger,
}

#[allow(dead_code)]
impl FileLogger {
    pub fn new(filename: &str) -> Self {
        let file = std::fs::OpenOptions::new()
            .create(true)
            .append(true)
            .open(filename)
            .unwrap();

        Self {
            logger: Logger {
                level: LogLevel::Info,
                use_colors: false,
                output: Mutex::new(Box::new(file)),
            },
        }
    }

    pub fn info(&self, message: &str) {
        self.logger.info(message);
    }

    pub fn error(&self, message: &str) {
        self.logger.error(message);
    }
}

#[allow(dead_code)]
pub struct MultiLogger {
    loggers: Vec<Logger>,
}

#[allow(dead_code)]
impl MultiLogger {
    pub fn new(loggers: Vec<Logger>) -> Self {
        Self { loggers }
    }

    pub fn debug(&self, message: &str) {
        for logger in &self.loggers {
            logger.debug(message);
        }
    }

    pub fn info(&self, message: &str) {
        for logger in &self.loggers {
            logger.info(message);
        }
    }

    pub fn warn(&self, message: &str) {
        for logger in &self.loggers {
            logger.warn(message);
        }
    }

    pub fn error(&self, message: &str) {
        for logger in &self.loggers {
            logger.error(message);
        }
    }

    pub fn fatal(&self, message: &str) {
        for logger in &self.loggers {
            logger.fatal(message);
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_log_levels() {
        let logger = Logger::new(LogLevel::Info);
        logger.debug("This should not appear");
        logger.info("This should appear");
        logger.warn("This should appear");
        logger.error("This should appear");
    }

    #[test]
    fn test_with_timing() {
        let logger = Logger::new(LogLevel::Info);
        let result = logger.with_timing("Test operation", || {
            std::thread::sleep(Duration::from_millis(10));
            42
        });
        assert_eq!(result, 42);
    }

    #[test]
    fn test_format_duration() {
        assert!(format_duration(Duration::from_millis(500)).contains("ms"));
        assert!(format_duration(Duration::from_secs(5)).contains("s"));
        assert!(format_duration(Duration::from_secs(90)).contains("m"));
    }
}
